youtube_dl/extractor/neteasemusic.py

```diff
@@ -35,32 +43,87 @@ def _encrypt(cls, dfsid):
+            'Referer': 'https://music.163.com',
+            'Cookie': cookie,
+        }
+        return ('params={}'.format(encrypted_params), headers)
+
+    @classmethod
+    def _call_player_api(cls, song_id, bitrate):
+        url = 'https://interface3.music.163.com/eapi/song/enhance/player/url'
+        data, headers = cls.make_player_api_request_data_and_headers(song_id, bitrate)
+        req = sanitized_Request(url, data=data.encode('ascii'), headers=headers)
+        try:
+            resp = compat_urllib_request.urlopen(req)
+        except Exception:
+            return {}
+        content = resp.read()
+        return json.loads(content)
+
     def extract_formats(self, info):
         formats = []
+        song_id = info['id']
         for song_format in self._FORMATS:
             details = info.get(song_format)
             if not details:
                 continue
-            song_file_path = '/%s/%s.%s' % (
-                self._encrypt(details['dfsId']), details['dfsId'], details['extension'])
-
-            # 203.130.59.9, 124.40.233.182, 115.231.74.139, etc is a reverse proxy-like feature
-            # from NetEase's CDN provider that can be used if m5.music.126.net does not
-            # work, especially for users outside of Mainland China
-            # via: https://github.com/JixunMoe/unblock-163/issues/3#issuecomment-163115880
-            for host in ('http://m5.music.126.net', 'http://115.231.74.139/m1.music.126.net',
-                         'http://124.40.233.182/m1.music.126.net', 'http://203.130.59.9/m1.music.126.net'):
-                song_url = host + song_file_path
-                if self._is_valid_url(song_url, info['id'], 'song'):
-                    formats.append({
-                        'url': song_url,
-                        'ext': details.get('extension'),
-                        'abr': float_or_none(details.get('bitrate'), scale=1000),
-                        'format_id': song_format,
-                        'filesize': details.get('size'),
-                        'asr': details.get('sr')
-                    })
-                    break
+
+            bitrate = details.get('bitrate', 999000)
+            data = self._call_player_api(song_id, bitrate)
+            if data and 'data' in data:
+                for song in data['data']:
+                    song_url = song['url']
+                    bitrate = song['br']
+                    size = song['size']
+
+                    if self._is_valid_url(song_url, info['id'], 'song'):
+                        formats.append({
+                            'url': song_url,
+                            'ext': details.get('extension'),
+                            'abr': float_or_none(bitrate, scale=1000),
+                            'format_id': song_format,
+                            'filesize': size,
+                            'asr': details.get('sr'),
+                        })
```

Comment:
Use utils.try_get() to simplify and increase robustness; also condition more values:
```suggestion
            for song in try_get(data, lambda x: x['data'], list) or []:
                song_url = try_get(song, lambda s: x['url'])
                if self._is_valid_url(song_url, info['id'], 'song'):
                    formats.append({
                        'url': song_url,
                        'ext': details.get('extension'),
                        'abr': float_or_none(song.get('br'), scale=1000),
                        'format_id': song_format,
                        'filesize': int_or_none(song.get('size')),
                        'asr': int_or_none(details.get('sr')),
                    })
```